src/backend/optimizer/path/allpaths.c

```diff
@@ -206,6 +207,36 @@ set_base_rel_pathlists(PlannerInfo *root)
 	}
 }
 
+
+/*
+ * handle_gen_olatile_path
+ *
+ * Change the path in its pathlist if match the pattern
+ * (segmentgeneral or general path contains volatile restrictions).
+ */
+static void
+handle_gen_volatile_path(PlannerInfo *root, RelOptInfo *rel)
+{
+	ListCell   *lc;
+
+	rel->cheapest_startup_path = NULL;
+	rel->cheapest_total_path = NULL;
+	rel->cheapest_unique_path = NULL;
+
+	foreach(lc, rel->pathlist)
+	{
+		Path	     *path = (Path *) lfirst(lc);
+
+		if (CdbPathLocus_IsGeneral(path->locus) &&
+			contain_volatile_functions((Node *) rel->baserestrictinfo))
+		{
+			CdbPathLocus_MakeSingleQE(&(path->locus));
```

Comment:
Yes, it is better to add a function call, because I am not sure that the order will not change in all cases. This is fixed in the latest fix.

Reply:
UPD: we can't call `add_path` twice for same path, because of [this Assert](https://github.com/greenplum-db/gpdb/blob/76dac6023ed7ffc211297a709fc82b0e94d623fb/src/backend/optimizer/util/pathnode.c#L934)

Reply:
I removed all `set_cheapest` call, because the cost of path doesn't depend on locus. At least I couldn't find any dependencies.